shhext/service.go

```diff
@@ -0,0 +1,145 @@
+package shhext
+
+import (
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+	whisper "github.com/ethereum/go-ethereum/whisper/whisperv6"
+)
+
+// ConfirmationHandler used as a callback for confirming that envelopes were sent.
+type ConfirmationHandler func(common.Hash)
+
+// Service is a service that provides some additional Whisper API.
+type Service struct {
+	w       *whisper.Whisper
+	tracker *tracker
+}
+
+// Make sure that Service implements node.Service interface.
+var _ node.Service = (*Service)(nil)
+
+// New returns a new Service.
+func New(handler ConfirmationHandler, w *whisper.Whisper) *Service {
+	track := &tracker{
+		w:       w,
+		handler: handler,
+		cache:   map[common.Hash]bool{},
+	}
+	return &Service{
+		w:       w,
+		tracker: track,
+	}
+}
+
+// Protocols returns a new protocols list. In this case, there are none.
+func (s *Service) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{}
+}
+
+// APIs returns a list of new APIs.
+func (s *Service) APIs() []rpc.API {
+	return []rpc.API{
+		{
+			Namespace: "shhext",
+			Version:   "1.0",
+			Service:   NewPublicAPI(s.w, s.tracker),
+			Public:    true,
+		},
+	}
+}
+
+// Start is run when a service is started.
+// It does nothing in this case but is required by `node.Service` interface.
+func (s *Service) Start(server *p2p.Server) error {
+	s.tracker.Start()
+	return nil
+}
+
+// Stop is run when a service is stopped.
+// It does nothing in this case but is required by `node.Service` interface.
+func (s *Service) Stop() error {
+	s.tracker.Stop()
+	return nil
+}
+
+// tracker responsible for processing events for envelopes that we are interested in
+// and calling specified handler.
+type tracker struct {
+	w       *whisper.Whisper
+	handler ConfirmationHandler
+
+	mu    sync.Mutex
+	cache map[common.Hash]bool
```

Comment:
Maybe something like this for self-documentary:
```go
type EnvelopeState int

const (
  EnvelopePosted EnvelopeState = iota
  EnvelopeSent
)

cache map[common.Hash]EnvelopeState
```

or a comment that `false` means that an envelope was posted but was not broadcasted yet.